### PR TITLE
INTYGFV-15007: Implementation of question Grund for medicinskt underlag.

### PR DESCRIPTION
**Pull request by ChrilleSigerhed**

### CODE REVIEW

File: fk/luae_na/src/main/java/se/inera/intyg/common/luae_na/v1/model/converter/certificate/question/QuestionUnderlag.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+                        .numberOfDays(LIMIT)
+                        .build(),
+                }
+            )
+            .build();
+    }
+
+    private static CertificateDataValueMedicalInvestigation getMedicalInvestigationValue(List<Underlag> underlag, int id) {
+        return CertificateDataValueMedicalInvestigation.builder()
+            .typeId(UNDERLAG_SVAR_JSON_ID_4 + "[" + id + "].typ")
```
Comment (IamFriberg) [inline fk/luae_na/src/main/java/se/inera/intyg/common/luae_na/v1/model/converter/certificate/question/QuestionUnderlag.java:142]: Should we refactor this to make it more clear to why we set typeId here and below we also use a different typeId as a type?

File: fk/fk-parent/src/main/java/se/inera/intyg/common/fkparent/model/converter/RespConstants.java [hunk 2]
```diff
@@ -79,9 +79,16 @@ public final class RespConstants {
 
     public static final String UNDERLAG_SVAR_ID_4 = "4";
     public static final String UNDERLAG_TYP_DELSVAR_ID_4 = "4.1";
+    public static final String UNDERLAG_QUESTION_ID_0 = "4.1.0";
```
Comment (mww-sopra-steria) [inline fk/fk-parent/src/main/java/se/inera/intyg/common/fkparent/model/converter/RespConstants.java:82]: Are these needed? UNDERLAG_QUESTION_ID_0, UNDERLAG_QUESTION_ID_1, UNDERLAG_QUESTION_ID_2.

File: support/src/main/java/se/inera/intyg/common/support/facade/model/config/CertificateDataConfigMedicalInvestigation.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    @Getter(onMethod = @__(@Override))
+    String text;
+    @Getter(onMethod = @__(@Override))
+    String description;
+    String typeText;
+    String dateText;
+    String informationSourceText;
+    String informationSourceDescription;
+    List<MedicalInvestigation> list;
+    List<CodeItem> typeOptions;
```
Comment (mww-sopra-steria) [inline support/src/main/java/se/inera/intyg/common/support/facade/model/config/CertificateDataConfigMedicalInvestigation.java:52]: Already commented.. but this should probably be part of MedicalInvestigation.

File: support/src/main/java/se/inera/intyg/common/support/facade/model/value/CertificateDataValueMedicalInvestigation.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
+import java.time.LocalDate;
+import lombok.Builder;
+import lombok.Value;
+import se.inera.intyg.common.support.facade.model.value.CertificateDataValueMedicalInvestigation.MedicalInvestigationValueBuilder;
+
+@JsonDeserialize(builder = MedicalInvestigationValueBuilder.class)
+@Value
+@Builder
+public class CertificateDataValueMedicalInvestigation {
```
Comment (mww-sopra-steria) [inline support/src/main/java/se/inera/intyg/common/support/facade/model/value/CertificateDataValueMedicalInvestigation.java]: This should consist of DateValue, CodeValue and TextValue.

File: fk/luae_na/src/main/java/se/inera/intyg/common/luae_na/v1/model/converter/certificate/question/QuestionUnderlag.java [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+                    CertificateDataValidationMaxDate.builder()
+                        .id(UNDERLAG_SVAR_JSON_ID_4 + "[2].datum")
+                        .numberOfDays(LIMIT)
+                        .build(),
+                }
+            )
+            .build();
+    }
+
+    private static CertificateDataValueMedicalInvestigation getMedicalInvestigationValue(List<Underlag> underlag, int id) {
```
Comment (mww-sopra-steria) [inline fk/luae_na/src/main/java/se/inera/intyg/common/luae_na/v1/model/converter/certificate/question/QuestionUnderlag.java:140]: See comment on CertificateDataValueMedicalInvestigation. This complex type should consist of DateValue, CodeValue and TextValue.

**Review by mww-sopra-steria — CHANGES_REQUESTED** [review]
A few things concerning the model that we need to revisit before we can complete this.

File: fk/fk-parent/src/main/java/se/inera/intyg/common/fkparent/model/converter/RespConstants.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (IamFriberg) [reply fk/fk-parent/src/main/java/se/inera/intyg/common/fkparent/model/converter/RespConstants.java:82]: These have been removed.

File: fk/fk-parent/src/main/java/se/inera/intyg/common/fkparent/model/internal/Underlag.java [hunk 6]
```diff
@@ -76,6 +75,11 @@ public String getLabel() {
             return label;
         }
 
+        public static UnderlagsTyp fromValue(String id) {
```
[the remark this answers is not in the document]
Comment (IamFriberg) [reply fk/fk-parent/src/main/java/se/inera/intyg/common/fkparent/model/internal/Underlag.java:78]: Removed

**Review by mww-sopra-steria — CHANGES_REQUESTED** [review]
Nice work! Just a small thing regarding the "withCitation" that would be great if you look at.

File: fk/luae_na/src/main/java/se/inera/intyg/common/luae_na/v1/model/converter/certificate/question/QuestionUnderlag.java [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+                        )
+                    )
+                    .build()
+            )
+            .validation(
+                new CertificateDataValidation[]{
+                    CertificateDataValidationMandatory.builder()
+                        .questionId(UNDERLAG_TYP_DELSVAR_ID_4)
+                        .expression(
+                            multipleAndExpressionWithCitation(UNDERLAG_SVAR_JSON_ID_4 + "[0].typ", UNDERLAG_SVAR_JSON_ID_4 + "[0].datum",
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply fk/luae_na/src/main/java/se/inera/intyg/common/luae_na/v1/model/converter/certificate/question/QuestionUnderlag.java:128]: Good suggestion, done!

File: support/src/main/java/se/inera/intyg/common/support/facade/util/ValidationExpressionToolkit.java [hunk 8]
```diff
@@ -58,6 +58,17 @@ public static String multipleAndExpression(String... expression) {
         });
     }
 
+    public static String multipleAndExpressionWithCitation(String... expression) {
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply support/src/main/java/se/inera/intyg/common/support/facade/util/ValidationExpressionToolkit.java:61]: I agree.

**Review by mww-sopra-steria — APPROVED** [review]
Looks good! Approved!

**Pull request merged**